ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp

```diff
@@ -122,33 +156,98 @@ size_t TensorLayout::compute_page_size_bytes(const Size& page_size) const {
+
+        // Iterate dims in reverse order and ensure alignment
+        // Even tensor of rank 0 or 1 must be aligned (to Tile / Page / Shard)
+        for (int i = -1; i >= -max_rank; --i) {
+            auto& dim = i == -1 ? width : height;
+            if(i >= -rank) {
+                dim *= shape[i];
+            }
+
+            // Align the current dimension if alignment is available
+            if (i >= -alignment_rank) {
+                dim = CMAKE_UNIQUE_NAMESPACE::round_up(dim, alignment_[i]);
+            }
         }
 
-        // Align the current dimension if alignment is available
-        if (i >= -alignment_rank) {
-            dim = CMAKE_UNIQUE_NAMESPACE::round_up(dim, alignment_[i]);
+        Size size{height, width};
+        return size;
+    } else {
+        // Iterate dims in reverse order
+        // Even tensor of rank 0 or 1
+        for (int i = -1; i >= -rank; --i) {
+            auto& dim = i == -1 ? width : height;
+            dim *= shape[i];
         }
-    }
 
-    Size size{height, width};
-    return size;
+        auto compute_default_logical_shard_shape = [&shape, &rank]() {
+            size_t shard_width = 1;
+            size_t shard_height = 1;
+
+            if (rank >= 1) {
+                shard_width = shape[-1];
+            }
+            if (rank >= 2) {
+                shard_height = shape[-2];
+            }
+
+            return Size{shard_height, shard_width};
+        };
+
+        const auto& logical_shard_shape = memory_config_.shard_spec.has_value() ? Size(memory_config_.shard_spec.value().shape) : compute_default_logical_shard_shape();
+        const auto& physical_shard_shape = compute_physical_shard_shape(logical_shard_shape);
+
+        auto get_physical_size = [](auto original_size, auto logical_shard_size, auto physical_shard_size, auto alignment) {
+            if (logical_shard_size == 0) {
+                return (uint32_t) 0;
+            }
+            // If we always pad to full shards, then return:
+            // auto num_shards = tt::div_up(original_size, logical_shard_size);
+            // return (uint32_t) physical_shard_size * num_shards;
+
+            // But host physical data is only padded up to nearest alignment (and no padding in between shards)
+            // So last shard is only up to nearest alignment
+            auto num_full_shards = original_size / logical_shard_size;
+            auto last_physical_shard_size = CMAKE_UNIQUE_NAMESPACE::round_up(original_size % logical_shard_size, alignment);
+            return (uint32_t) (physical_shard_size * num_full_shards + last_physical_shard_size);
+        };
+
+        auto physical_height = get_physical_size(height, logical_shard_shape.height(), physical_shard_shape.height(), alignment_rank == 2 ? alignment_[-2] : 0);
+        auto physical_width = get_physical_size(width, logical_shard_shape.width(), physical_shard_shape.width(), alignment_rank >= 1 ? alignment_[-1] : 0);
```

Comment:
if you are ok that it will be padded up to 1, not 0, you can remove rank checks. alignment is using the VectorBase which supports negative indexing of up to -4
```suggestion
        auto physical_height = get_physical_size(height, logical_shard_shape.height(), physical_shard_shape.height(), alignment_[-2]);
        auto physical_width = get_physical_size(width, logical_shard_shape.width(), physical_shard_shape.width(), alignment_[-1]);
```